components/repo/Finder.tsx

```diff
@@ -0,0 +1,150 @@
+import { useState, useEffect } from 'react'
+import { Combobox } from '@headlessui/react'
+import client from '@/lib/client'
+import { classNames } from '@/lib/util'
+
+type TypeaheadResult = {
+  did: string
+  handle: string
+  avatar?: string
+  displayName?: string
+}
+
+type RepoFinderProps = {
+  selectionType?: 'did' | 'handle'
+  onChange: (value: string) => void
+  inputProps?: React.InputHTMLAttributes<HTMLInputElement>
+}
+
+const DefaultResult = {
+  did: '',
+  handle: 'DID or Handle',
+  displayName: 'Start typing to search...',
+}
+
+const ErrorResult = {
+  did: '',
+  handle: 'Error searching',
+  displayName: 'Start typing to search...',
+}
+
+const getProfilesForQuery = async (q: string): Promise<TypeaheadResult[]> => {
+  const headers = { headers: client.proxyHeaders() }
+  if (q.startsWith('did:')) {
+    const { data: profile } = await client.api.app.bsky.actor.getProfile(
+      {
+        actor: q,
+      },
+      headers,
+    )
+
+    return profile
+      ? [
+          {
+            displayName: profile.displayName,
+            handle: profile.handle,
+            avatar: profile.avatar,
+            did: profile.did,
+          },
+        ]
+      : []
+  }
+
+  const {
+    data: { actors },
+  } = await client.api.app.bsky.actor.searchActorsTypeahead({ q }, headers)
+
+  return actors.map((actor) => ({
+    displayName: actor.displayName,
+    handle: actor.handle,
+    avatar: actor.avatar,
+    did: actor.did,
+  }))
+}
+
+export function RepoFinder({
+  selectionType = 'did',
+  onChange,
+  inputProps = {},
+}: RepoFinderProps) {
+  const [query, setQuery] = useState('')
+  const [selectedItem, setSelectedItem] = useState<string>('')
+  const [items, setItems] = useState<TypeaheadResult[]>([DefaultResult])
+  const [loading, setLoading] = useState(false)
+
+  useEffect(() => {
+    if (query.length > 0) {
+      setLoading(true)
+      getProfilesForQuery(query)
+        .then((profiles) => {
+          setItems(profiles)
+          setLoading(false)
+        })
+        .catch((error) => {
+          console.error('Error fetching data:', error)
+          setLoading(false)
+          setItems([ErrorResult])
+        })
+    } else {
+      setItems([DefaultResult])
+    }
+  }, [query])
+
+  return (
+    <Combobox
+      value={selectedItem}
+      onChange={(item) => {
+        setSelectedItem(item)
+        onChange(item)
+      }}
```

Comment:
It is a little complicated to have both internal state and external state working at the same time, as it means consumers may be prone to get out of sync with the internal state of the component.  Would it be awkward to make `Finder` a standard controlled form component?

Reply:
surely doable but only issue is that in a lot of places, we use uncontrolled inputs and then utilize web native html form methods to get data. I like that pattern for the most part and only use controlled method when absolutely needed so I left the default behavior to match that.

will do a refactor if we end up seeing issues with this.